server/src/main/java/org/elasticsearch/health/node/DiskHealthIndicatorService.java

```diff
@@ -0,0 +1,284 @@
+                            " As a result %d ind%s cannot process any more updates.",
+                            problemIndices.size(),
+                            problemIndices.size() > 1 ? "ices" : "ex"
+                        )
+                        : ""
+                );
+                if (problemIndices.size() > 1) {
+                    impacts.add(RED_HEALTH_AFFECTED_INDICES_INGEST_IMPACT);
+                }
+            } else {
+                problemNodes = diskHealthInfoMap.entrySet()
+                    .stream()
+                    .filter(entry -> HealthStatus.YELLOW.equals(entry.getValue().healthStatus()))
+                    .map(Map.Entry::getKey)
+                    .collect(Collectors.toSet());
+                problemIndices = getIndicesForNodes(problemNodes);
+                symptom = String.format(
+                    Locale.ROOT,
+                    "%d node%s increased disk space.%s",
+                    problemNodes.size(),
+                    problemNodes.size() > 1 ? "s have" : " has",
+                    problemIndices.size() > 0
+                        ? String.format(
+                            Locale.ROOT,
+                            " As a result %d ind%s at risk of not being able to process any more updates.",
+                            problemIndices.size(),
+                            problemIndices.size() > 1 ? "ices are" : "ex is"
+                        )
+                        : ""
+                );
+                if (problemIndices.size() > 1) {
+                    impacts.add(YELLOW_HEALTH_AFFECTED_INDICES_INGEST_IMPACT);
+                }
+            }
+            Set<DiscoveryNodeRole> roles = clusterService.state()
+                .nodes()
+                .getNodes()
+                .values()
+                .stream()
+                .filter(node -> problemNodes.contains(node.getId()))
+                .map(DiscoveryNode::getRoles)
+                .flatMap(Collection::stream)
+                .collect(Collectors.toSet());
+            if (roles.contains(DiscoveryNodeRole.MASTER_ROLE)) {
+                impacts.addAll(StableMasterHealthIndicatorService.UNSTABLE_MASTER_IMPACTS);
+            }
+            if (roles.contains(DiscoveryNodeRole.INGEST_ROLE)) {
+                impacts.add(HEALTH_INGEST_NODES_PIPELINES_IMPACT);
+                impacts.add(HEALTH_INGEST_NODES_MONITORING_IMPACT);
+            }
+            if (roles.contains(DiscoveryNodeRole.ML_ROLE)) {
+                impacts.add(HEALTH_ML_NODES_IMPACT);
+            }
+            if (roles.contains(DiscoveryNodeRole.REMOTE_CLUSTER_CLIENT_ROLE)) {
+                impacts.add(HEALTH_REMOTE_CLUSTER_CLIENT_NODES_IMPACT);
+            }
+            if (roles.contains(DiscoveryNodeRole.TRANSFORM_ROLE)) {
+                impacts.add(HEALTH_TRANSFORM_NODES_IMPACT);
+            }
+            diagnosisList = getDiskProblemsDiagnosis(explain, problemNodes, problemIndices);
```

Comment:
I believe the diagnosis will be based on the node role right? 

Or is the plan to have one diagnosis and mention the various roles in the troubleshooting guide? 

IMO we should have a diagnosis for data nodes (with its own id and troubleshooting guide), one for master nodes, etc

What do you think? (@gmarouli )

Reply:
I think this requires some thinking. I am pretty confident that we need a distinction id a node contains data or not, because in the on-prem case adding nodes is not going to fix anything. 

We could start with this separation and see how it forms, what do you think?

About the affected resources, I agree I think it's time, for now I would stick with indices and nodes, but I think we should abstract maybe data streams, tiers etc. Any thoughts?

Reply:
> Or is the plan to have one diagnosis and mention the various roles in the troubleshooting guide?

Based on the spreadsheet we've been working from I had been thinking one diagnosis for all roles (since in all cases I think it's "find the node and clean up space or add disk") but different impacts for each role (since that will truly be different). We can make different diagnoses for each role but we'll need to work out what those would be.

Reply:
> I am pretty confident that we need a distinction id a node contains data or not, because in the on-prem case adding nodes is not going to fix anything.

That's the one thing we do have here right? If a node has an index, we give a data-specific diagnosis and impact. If the problematic nodes don't have index data then we don't recommend adding nodes. Or am I misunderstanding?

Reply:
> Also, is it time we add a type to the affected_resource and report the problematic node_ids there?

> About the affected resources, I agree I think it's time, for now I would stick with indices and nodes

Makes sense. I started doing it though and it is a fairly wide change (simple, but touches a lot). And so far I've run across resource types index, node, policy, and repository. I think we ought to do this as a separate PR (which could be before or after this one) so that we don't wind up making this PR several times bigger.

Reply:
I pushed that code into a separate branch so that we could still talk about it but not have it in this PR: https://github.com/elastic/elasticsearch/pull/90077. I'm not tied to that at all if someone wants to take it in a different direction, but wanted to give an idea of the scope of the change.

Reply:
> Based on the spreadsheet we've been working from I had been thinking one diagnosis for all roles (since in all cases I think it's "find the node and clean up space or add disk") 

@masseyke I think the solutions **will** be different. ie. enable autoscaling for data nodes, add capacity for master nodes

Reply:
>I pushed that code into a separate branch so that we could still talk about it but not have it in this PR: https://github.com/elastic/elasticsearch/pull/90077. I'm not tied to that at all if someone wants to take it in a different direction, but wanted to give an idea of the scope of the change.

Thanks Keith. I think we'll want something along the lines of
```
“affected_resources”: { 
  { “type”: “index”, “resources” : [ “index1”, “index2”, …] },
  {  “type”: “node”, “resources” : [ “node-1”, “node-2, …] }, 
}
```
or
```
“affected_resources”: { 
  “type”: “index”, “resources” : [ “index1”, “index2”, …] ,
}
```
But let's have that discussion in a different medium

For the scope of this PR shall we report the node ids in the `affected_resources` ?

Reply:
> For the scope of this PR shall we report the node ids in the affected_resources ?

Yeah that's what I'm proposing.

Reply:
> @masseyke I think the solutions will be different. ie. enable autoscaling for data nodes, add capacity for master nodes

And that's in line with what we have here right? Different diagnoses for data nodes vs other nodes. I think I'm missing something.